Dynatrace/dynatrace-operator
# Support for IBM Power and fix proxy bug affecting cloud-native app-only monitoring

# Description

EDIT: Closed in favor of #865 and #866 

This PR brings application-only monitoring support to OpenShift on IBM Power and Z.

Z-support is affected by a bug where `ruxitagentproc.conf` is missing from the oneagent install.zip for the s390 platform.

The motivation behind this is of joint concern to Dynatrace, IBM and Red Hat, to target a growing customer audience that is requesting Dynatrace Operator on the IBM Power and Z platforms.

## How can this be tested?
### New Build Scripts Provided
To get the operator code to build (with CGO enabled) for Power and Z, it was required to build locally against a RHEL 8 host. I would prefer a container/UBI-8 based build solution, but that was not possible due to the lack of a UBI-based go-toolset v1.17+ when I started this back in January. There is a new RHEL-specific container build script (a modified version of `./build/deploy_local.sh`) provided for this, as well as a script for building the bundle and index images, as well as the catalog source yaml file for deploying the index onto an OpenShift cluster, for installation using OperatorHub which is required for Red Hat certification of the operator.

### Requirements to Build and Test
To test this yourself, you'd need the following:
* A RHEL 8 VM running natively on Power or Z in order to build the operator (native container-based builds are now possible but not provided with this PR) with the following installed
   * Go 1.18 installed from go.dev
   * The following additional RHEL packages: `git gcc gcc-c++ make device-mapper-devel golang podman podman-docker gpgme-devel libassuan-devel`
   * Operator SDK v1.16.0
   * Operator Registry (`opm`) v1.19.5
   * The following go binaries:
      * sigs.k8s.io/controller-tools/cmd/controller-gen@v0.3.0
      * sigs.k8s.io/kustomize/kustomize/v3
   * Helm v3
* Power or Z-based OpenShift cluster, to be able to test app-only monitoring on the respective platform

### Steps to Build:

With the above installed and running on the local RHEL 8 VM, the `deploy_rhel.sh` script is called as follows:

```
$ GOARCH=<s390x|ppc64le> LOCALBUILD=1 \
 IMG=quay.io/orgname/dynatrace-operator TAG=v0.6.0-<arch> \
./hack/build/deploy_rhel.sh
```
You should now have the operator image built and pushed to the quay repository specified, with the tag that was specified.

Next, prepare the bundle, index and catalog source for deployment onto OpenShift with the `bundle_deploy_ocp.sh` script:

```
$ ./hack/build/bundle_deploy_ocp.sh 0.6.0-<arch> <quay-orgname>
```

The above will create and push a bundle image, followed by an index image, to the `<quay-orgname>` with the repository of `dynatrace-operator-bundle:v0.6.0-<arch>` and `operator-index:v<timestamp>-<build#>`, respectively. You will also see a catalog source yaml file named <quay-orgname>-operators.catalogsource.yaml`, with the respective bundle image and `<quay-orgname>` specified within the catalog source.

To deploy the operator bundle on an OpenShift cluster, first login to OpenShift using `oc login <api-url>` and simply apply the catalog source:

```
oc create -f <quay-orgname>-operators.catalogsource.yaml
```

### Steps to Test
In order to test, install the operator from the web console (I have a script, also) into the `openshift-operators` (All Operators) namespace.

For an application to monitor, I install the IBM OpenLiberty v0.8.1 operator from here (requires building the application container on the specific platform): https://github.com/OpenShift-Z/openliberty-operator-ocpz 

I create the required secret, containing the dynatrace PaaS and API Tokens/Keys into the `openshift-operators` namespace.

By default, that application runs in the `openliberty-demo` namespace. I label this namespace with the label `monitor: applicationMonitoring` and deploy the following CR into the `openshift-operators` namespace:

```
apiVersion: dynatrace.com/v1beta1
kind: DynaKube
metadata:
  name: dynakube
  namespace: openshift-operators
spec:
  apiUrl: https://<uuid>.live.dynatrace.com/api

  namespaceSelector:
    matchLabels:
      monitor: applicationMonitoring

  oneAgent:
      useCSIDriver: false
```

You should then see application metrics visible in your Dynatrace Dashboard, for the specified live environment `<uuid>`.

Please ping me over slack or via this PR for further questions and discussion.

## Checklist
